### PR TITLE
Editorial: Simplify operations (#2248)

### PR DESCRIPTION
**Pull request by 12wrigja**
Similar to other rebases, I intend for this to be merged as-is without squashing.

The typing for GetTemporalUnit is interesting - ideally, I'd like to find a way to avoid all the casts within the function itself and avoid needing overloads, but I can't seem to get TS to recognize that when the default empty array (typed as `never[]`) is used as the `extraValues` property that it expands to `never` in the return type (which is the union identity).

### CODE REVIEW

File: lib/ecmascript.ts
Comment (ptomato) [inline lib/ecmascript.ts:871]: Not sure I get why this is necessary from just reading the comment and looking at the signature of the first overload of GetTemporalUnit. I'd expect the return type of that overload to be `Temporal.TimeUnit | 'auto' | undefined` which seems like it shouldn't make the condition impossible.

File: lib/ecmascript.ts
Comment (ptomato) [inline lib/ecmascript.ts:943]: If I'm reading this correctly, `Exclude<D, typeof REQUIRED>` already is unioned with `UnitTypeMapping[T]`, so that might be redundant?

**Review by ptomato — APPROVED** [review]
Looks good, just remarking on some oddities that I noticed which may be just my inexperience with TS.

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts]: This is leftover from a previous iteration of the GetTemporalUnit overloads which I screwed up. Thanks for catching - you are right about the return type and this not needing the cast.

File: lib/ecmascript.ts
Comment (12wrigja) [reply lib/ecmascript.ts:939]: I don't believe it's redundant because of the way that the binding D happens.

If I write

`GetTemporalUnit(o, 'smallestUnit', 'time', 'hours')`

then `D` is bound to `'hours'`, not to the whole union type space.

Therefore, we need to re-add that space back into the return type (as the allowed values also include all `TimeUnit` values).

**Comment by justingrant** [comment]
Sorry I'm coming late to this party!  I had it starred in my email app but had a busy few weeks and didn't get to it until now. 

The typing of these Largest/Smallest unit methods was one of the 2 hardest TS challenges when I originally did the TS conversion. The combination of allowing and disallowing options was painful to make typed. 

I have one question on the new implementation. If I remember correctly, some of the methods allowed `auto` but others didn't.  Has that distinction been retained in the new TS types, so we'll catch the case of someone passing `auto` to a method that shouldn't allow it?

**Comment by 12wrigja** [comment]
> I had it starred in my email app

In the future, feel free to drop a line here saying something like "id like to review this before it's merged" and I'm happy to wait for your review.

> some of the methods allowed

I think you are referring to `ToSmallestTemporalUnit` didn't accept `auto`, but `ToLargestTemporalUnit` did. I didn't notice that when I was merging this PR, and dropped that detail. There's another commit upstream that further modifies the new impl signature, so I'll incorporate this change into that PR.

**Pull request merged**
